yetanotherco/aligned_layer
# feat(cli): Price Estimates in Aligned CLI

# Price Estimates in Aligned CLI

## Description

Addresses No. 1 in #1570. Adds a new pricing parameter for the cli using `PriceEstimate` from the sdk. Refactors some parts of the sdk.

## Type of change

Please delete options that are not relevant.


### To Test:
- Start local devnet
- If you prefer to use a paying address run the following: 
```
cast send --rpc-url http://localhost:8545 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 --value 9999.9999999ether --private-key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```
- Try submitting proofs with the following commands and observe the following edge cases are successfully submit a proof to the batcher:

Not specifying a fee should use the default.
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```

<img width="740" alt="Screenshot 2024-12-05 at 20 30 24" src="https://github.com/user-attachments/assets/b4346724-7603-4af5-9f74-405bea744a14">

`--max_fee`
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --max_fee 0.001ether --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```

`--default_fee_estimate`
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --default_fee_estimate --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```

`--instant_fee_estimate`
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --instant_fee_estimate --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```

`--custom_fee_estimate <NUM_PROOFS_IN_BATCH>"
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --custom_fee_estimate <NUM_PROOFS_IN_BATCH> --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```

- Try submitting proofs with the following commands and observe the following edge cases are triggered:

`max_fee` and `default_fee_estimate/instant_fee_estimate/custom_fee_estimate` should not be displayed.
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --max_fee 1000000000 --custom_fee_estimate --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```
<img width="880" alt="Screenshot 2024-12-05 at 20 28 45" src="https://github.com/user-attachments/assets/d9eab46b-6463-4919-905d-168b19484cc5">


Clap errors if custom does not have <NUM_PROOFS_PER_BATCH>
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --custom_fee_estimate --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```
<img width="1028" alt="Screenshot 2024-12-05 at 20 29 20" src="https://github.com/user-attachments/assets/a27ea1e0-287c-46e3-a704-367f4941290a">


Parsing errors for "--custom_fee_estimate" if incorrect value is supplied for `num_proofs_in_batch`
```
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 20 \
                --proof_generator_addr 0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65 \
                --rpc_url http://localhost:8545 \
                --network devnet --price_estimate "custom u" --private_key 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a
```
<img width="1028" alt="Screenshot 2024-12-05 at 20 29 44" src="https://github.com/user-attachments/assets/e206fc09-be89-45f7-9457-efbd40b6dc78">



## Checklist

  - [Link to the open issue addressing it]()
